generators/protocol_block_ids.py

```diff
@@ -0,0 +1,275 @@
+            data += "return " + str(self.states[0]["id"]) + ";\n"
+        data += "}\n"
+
+        return data
+
+    def paletteToProtocol(self):
+        data = "constexpr BlockId paletteToProtocol(std::vector<std::pair<std::string, std::string>> properties) {\n"
+        data += "if (properties.size() != " + str(len(self.properties)) + ")\n"
+        data += "throw std::runtime_error(\"Invalid number of properties\");\n"
+        if self.properties != []:
+            for prop in self.properties:
+                data += "Properties::" + prop.capitalize() + " " + prop + ";\n"
+            data += "for (auto prop : properties) {\n"
+            first_prop = True
+            for prop in self.properties:
+                first_value = True
+                if first_prop:
+                    data += "if (prop.first == \"" + prop + "\") {\n"
+                    first_prop = False
+                else:
+                    data += "} else if (prop.first == \"" + prop + "\") {\n"
+                for each in self.properties[prop]:
+                    if first_value:
+                        data += "if (prop.second == \"" + each + "\") {\n"
+                        first_value = False
+                    else:
+                        data += "} else if (prop.second == \"" + each + "\") {\n"
+                    data += prop + " = Properties::" + prop.capitalize() + "::" + (num2words(each).replace("-", "_").upper() if each.isdigit() else each.upper()) + ";\n"
+                data += "} else {\n"
+                data += "throw std::runtime_error(\"Invalid property \\\"" + prop + "\\\" value\");\n"
+                data += "}\n"
+            data += "} else {\n"
+            data += "throw std::runtime_error(\"Invalid property name\");\n"
+            data += "}\n"
+            data += "}\n"
+            data += "return toProtocol(" + ", ".join(self.properties) + ");\n"
+        else:
+            data += "return toProtocol();\n"
+        data += "}\n"
+        return data
+
+    def namespace(self):
+        data = "namespace " + self.name.split(":")[1].title().replace("_", "") + " {\n"
+        data += self.Properties()
+        data += self.toProtocol()
+        data += self.paletteToProtocol()
+        data += "}\n"
+        return data + "\n"
+
+    def nameToProtocolId(self):
+        return "{\"" + self.name + "\", " + self.name.split(":")[1].title().replace("_", "") + "::paletteToProtocol},\n"
+
+    def toName(self):
+        global number_of_protocol_ids
+        data = ""
+        for state in self.states:
+            number_of_protocol_ids += 1
+            data += "case " + str(state["id"]) + ":\n"
+            data += "return {\"" + self.name + "\", {"
+            if self.properties != []:
+                for prop in self.properties:
+                    data += "{\"" + prop + "\", \"" + state["properties"][prop] + "\"}, "
+                data = data[:-2]
+            data += "}};\n"
+        return data
+
+blocks = []
```

Comment:
Put theses in a function, because this script could be called from another one idk.

Reply:
call it from `if __name__ == "__main__":` thingy